### PR TITLE
#1069: Added support for writing ttnn and ttmetal ops through python infrastructure and lowering it to flatbuffer files

### PR DESCRIPTION
**Pull request by tapspatel**
This PR introduces enhanced support from @kmitrovicTT 's python TTIR lowering into flatbuffer infra. It is a precursor to embedded golden functionality.

Blocked by: https://github.com/tenstorrent/tt-mlir/pull/1089

### CODE REVIEW

**Comment by tapspatel** [comment]
Building off of the awesome work done by @kmitrovicTT :D

File: lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ void createTTNNPipelineTTIRPasses(
 
   // Inlines all private functions. I.e flattens the program into the main
   // function. Removes all private functions.
-  pm.addPass(mlir::createInlinerPass());
+  // pm.addPass(mlir::createInlinerPass());
```
Comment (tapspatel) [inline lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:26]: @LPanosTT I was having trouble with this pass. Got the following error
```
LLVM ERROR: checking for an interface (mlir::DialectInlinerInterface) that was promised by dialect 'func' but never implemented
```

Is this pass supposed to get registered somewhere? I don't see it in the tablegen files. Only reference I see to it is TTIRInlinerInterface.

Comment (LPanosTT) [reply lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:26]: You’ll need to call mlir::tt::registerAllExtensions() and mlir::func::registerAllExtensions() before you run your passes

**Comment by tapspatel** [comment]
@nsmithtt is there some clean way we can run python files through llvm-lit config? I was experimenting with adding `.py` in `config.suffixes` with minimal success. I have no idea how the tests under `test/python` are running. My guess is they are not

Ideally I want to add this into CI as well since this is definitely a user supported way to write .mlir files and generate flatbuffers

File: lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp (hunk 1 above)
Comment (tapspatel) [reply lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:26]: fantastic, mlir::tt::registerAllExtensions() fixed it. Thank you

File: test/python/test_ttir_ops_ttnn.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    translate_ttmetal_to_flatbuffer,
+    ttir_to_ttmetal,
+)
+from ttmlir.ttir_builder import Operand, TTIRBuilder
+
+
+@translate_ttnn_to_flatbuffer(output_file_name="test_exp.ttnn")
+@ttir_to_ttnn(output_file_name="test_exp.mlir", system_desc_path=f"{system_desc_path}")
+@compile_as_mlir_module((128, 128))
+def test_exp_ttnn(in0: Operand, builder: TTIRBuilder):
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply test/python/test_ttir_ops_ttnn.py:25]: yeah I agree. I went through all our metal tests and only have the ones that work - there isn't a 1:1 success rate yet.

File: python/test_infra/ttir_builder.py [hunk 3]
```diff
@@ -231,49 +235,38 @@ def empty(
...
 
     # ----- TTIR op factories -----
-
-    def add(self, in0: Operand, in1: Operand) -> OpView:
-        """Convenience wrapper constructing `ttir.AddOp`."""
-        assert self.get_shape(in0) == self.get_shape(
-            in1
-        ), "Elementwise `ttir.add` op expects inputs of same shape."
-
+    def unary_proxy(self, op_golden_function, op_ttir_function, in0: Operand) -> OpView:
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply python/test_infra/ttir_builder.py]: fixed

File: test/python/test_ttir_ops_ttmetal.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    translate_ttmetal_to_flatbuffer,
+    ttir_to_ttmetal,
+)
+from ttmlir.ttir_builder import Operand, TTIRBuilder
+
+
+@translate_ttmetal_to_flatbuffer(output_file_name="test_exp.ttm")
+@ttir_to_ttmetal(
+    output_file_name="test_exp.mlir", system_desc_path=f"{system_desc_path}"
+)
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply test/python/test_ttir_ops_ttmetal.py:25]: added cleanup tasks to follow-up for this issue

File: python/test_infra/ttir_builder.py [hunk 5]
```diff
@@ -231,80 +235,194 @@ def empty(
...
+        return self.eltwise_proxy(torch.lt, ttir.LessThanOp, [in0, in1])
+
+    def div(self, in0: Operand, in1: Operand) -> OpView:
+        return self.eltwise_proxy(torch.div, ttir.DivOp, [in0, in1])
+
+    def maximum(self, in0: Operand, in1: Operand) -> OpView:
+        return self.eltwise_proxy(torch.maximum, ttir.MaximumOp, [in0, in1])
+
+
+def compile_as_mlir_module(
```
Comment (rpavlovicTT) [inline python/test_infra/ttir_builder.py:328]: Any reason to have this decorator defined in 2 files? Should we remove one of them?

Comment (tapspatel) [reply python/test_infra/ttir_builder.py:328]: will remove, thanks for the catch

**Pull request merged**